src/Microsoft.AspNetCore.Authentication.Cookies/Events/CookieRedirectContext.cs

```diff
@@ -20,15 +20,25 @@ public class CookieRedirectContext : BaseCookieContext
         /// <param name="options">The cookie handler options</param>
         /// <param name="redirectUri">The initial redirect URI</param>
         /// <param name="properties">The <see cref="AuthenticationProperties"/>.</param>
-        public CookieRedirectContext(HttpContext context, AuthenticationScheme scheme, CookieAuthenticationOptions options, string redirectUri, AuthenticationProperties properties)
-            : base(context, scheme, options, properties)
+        public CookieRedirectContext(
+            HttpContext context,
+            AuthenticationScheme scheme,
+            CookieAuthenticationOptions options,
+            AuthenticationProperties properties,
+            string redirectUri)
+            : base(context, scheme, options)
         {
+            Properties = properties;
             RedirectUri = redirectUri;
         }
 
         /// <summary>
         /// Gets or Sets the URI used for the redirect operation.
         /// </summary>
         public string RedirectUri { get; set; }
+
+        public bool Skipped { get; private set; }
```

Comment:
> Can't they just override the logic? I thought that's the whole point of the events
 
For every event except the `OnRedirectToXYZ` events, overriding the logic means two cumulative things:

  - Telling the handler that you don't want it to apply the default logic (in this case, setting the `Location` response header and returning a 302 response if the request is not an AJAX request). That's exactly what `Skip()` is made for.

  - Providing the custom logic.

Unfortunately, the `OnRedirectToXYZ` events are totally inconsistent as they are the only ones that have a non-empty `OnRedirectToXYZ` property (the redirection logic is not in the scheme handler but in the property). I plan to fix that by moving the logic to the scheme handler so these events are consistent with the rest of the stack.

This way, if you only want to update the return URL, you'll do something like that:

```csharp
options.Events.OnRedirectToReturnUrl = context =>
{
    // Simply replace the target URL but don't override the default logic.
    context.RedirectUri = "http://www.google.com/";

    return Task.CompletedTask;
};
```

... and to completely override the default logic:

```csharp
options.Events.OnRedirectToReturnUrl = async context =>
{
    // Tell the handler you don't want to use the default logic.
    context.Skip();

    // Apply the custom logic.
    await context.Response.WriteAsync("Hello World");
};
```